pkg/storage/segments/store.go

```diff
@@ -46,16 +46,95 @@ type Store interface {
 }
 
 type segmentStore struct {
-	oc  overlay.Client
-	ec  ecclient.Client
-	pdb pointerdb.Client
-	rs  eestream.RedundancyStrategy
+	oc            overlay.Client
+	ec            ecclient.Client
+	pdb           pointerdb.Client
+	rs            eestream.RedundancyStrategy
+	thresholdSize int
 }
 
 // NewSegmentStore creates a new instance of segmentStore
 func NewSegmentStore(oc overlay.Client, ec ecclient.Client,
-	pdb pointerdb.Client, rs eestream.RedundancyStrategy) Store {
-	return &segmentStore{oc: oc, ec: ec, pdb: pdb, rs: rs}
+	pdb pointerdb.Client, rs eestream.RedundancyStrategy, t int) Store {
+	return &segmentStore{oc: oc, ec: ec, pdb: pdb, rs: rs, thresholdSize: t}
+}
+
+// PeekThresholdReader allows a check to see if the size of a given reader
+// exceeds the maximum inline segment size or not.
+type PeekThresholdReader struct {
+	r              io.Reader
+	n              int // number of bytes read into thresholdBuf
+	thresholdBuf   []byte
+	totalReadBytes int
+}
+
+// NewPeekThresholdReader creates a new instance of PeekThresholdReader
+func NewPeekThresholdReader(r io.Reader) (pt *PeekThresholdReader) {
+	return &PeekThresholdReader{r: r, n: 0, thresholdBuf: nil, totalReadBytes: 0}
+}
+
+// Read initially reads bytes from the internal buffer, then continues
+// reading from the wrapped data reader. The number of bytes read `n`
+// is returned.
+func (pt *PeekThresholdReader) Read(p []byte) (n int, err error) {
+
+	// Case 1: if the total number of bytes read is greater than the number
+	// of bytes read into the thresholdBuf, then Read is called on the given
+	// byte slice.
+	if pt.totalReadBytes > pt.n {
+		return pt.r.Read(p)
+	}
+
+	thresholdBytesRemaining := pt.n - pt.totalReadBytes
+
+	// Case 2: if the length of the given byte slice p is less than or equal
+	// to the number of threshold bytes remaining to be read, then the slice
+	// of the threshold buffer from the end of totalReadBytes to the length
+	// of p is copied to p.
```

Comment:
If we only had `thresholdBuf` as suggested above, the logic for cases 2 and 3 would be:
```
n := copy(p, thresholdBuf)
thresholdBuf = tresholdBuf[n:]
return n, nil
```